egs/librispeech/ASR/prepare_gigaspeech.py

```diff
@@ -0,0 +1,293 @@
+        "If it's larger than 0, determines in which direction (relative to the supervision) "
+        "to seek for extra acoustic context. Available values: (left|right|center|random).",
+    )
+    parser.add_argument(
+        "--precomputed-features",
+        type=str2bool,
+        default=True,
+        help="Should we pre-compute features and store them on disk or not. "
+        "It is recommended to disable it for L and XL splits as the pre-computation "
+        "might currently consume excessive memory and time -- use on-the-fly feature "
+        "extraction in the training script instead.",
+    )
+    return parser
+
+
+# Similar text filtering and normalization procedure as in:
+# https://github.com/SpeechColab/GigaSpeech/blob/main/toolkits/kaldi/gigaspeech_data_prep.sh
+
+
+def normalize_text(
+    utt: str,
+    punct_pattern=re.compile(r"<(COMMA|PERIOD|QUESTIONMARK|EXCLAMATIONPOINT)>"),
+    whitespace_pattern=re.compile(r"\s\s+"),
+) -> str:
+    return whitespace_pattern.sub(" ", punct_pattern.sub("", utt))
+
+
+def has_no_oov(
+    sup: SupervisionSegment, oov_pattern=re.compile(r"<(SIL|MUSIC|NOISE|OTHER)>")
+) -> bool:
+    return oov_pattern.search(sup.text) is None
+
+
+def main():
+    args = get_parser().parse_args()
+    dataset_parts = [args.subset, "DEV", "TEST"]
+    if args.subset in ["L", "XL"]:
+        assert is_module_available("pyarrow"), (
+            "Running the GigaSpeech recipe for L and XL splits "
+            "currently requires installing optional dependencies: "
+            "'pip install pyarrow pandas'."
+        )
+
+    print("Parts we will prepare: ", dataset_parts)
+
+    corpus_dir = locate_corpus(
+        Path("/export/corpora5/gigaspeech"),
+        Path("/exp/pzelasko/gigaspeech"),
+        Path("/home/storage07/zhangjunbo/data/GigaSpeech")
+    )
+    musan_dir = locate_corpus(
+        Path("/export/corpora5/JHU/musan"),
+        Path("/export/common/data/corpora/MUSAN/musan"),
+        Path("/root/fangjun/data/musan"),
+    )
+
+    output_dir = Path("exp/data")
+    print("GigaSpeech manifest preparation:")
+    gigaspeech_manifests = prepare_gigaspeech(
+        corpus_dir=corpus_dir,
+        dataset_parts=dataset_parts,
+        output_dir=output_dir,
+        num_jobs=args.num_jobs,
+    )
+
+    print("Musan manifest preparation:")
+    musan_cuts_path = output_dir / "cuts_musan.json.gz"
+    musan_manifests = prepare_musan(
+        corpus_dir=musan_dir, output_dir=output_dir, parts=("music", "speech", "noise")
+    )
+
+    ctx_suffix = get_context_suffix(args)
+
+    print("Feature extraction:")
+    extractor = Fbank(FbankConfig(num_mel_bins=80))
+    with get_executor() as ex:  # Initialize the executor only once.
+        for partition, manifests in gigaspeech_manifests.items():
+            # For L and XL partition we are going to store the manifest using pyarrow.
+            cuts_path_ext = "jsonl.gz" if partition not in ["L", "XL"] else "arrow"
```

Comment:
This is also from an older version, please try to replace this with the recipe code on current master in snowfall.

BTW I originally skipped feature extraction for L and XL because it was too slow and took too much memory. But since then @csukuangfj added chunked feature extraction in kaldifeat, and I implemented CUDA extraction in Lhotse. You can safely compute features for everything with the new mechanisms. See the following method on CutSet with examples:

https://github.com/lhotse-speech/lhotse/blob/ce1437c8127d3545f7c45a5677c44bbb5b8d4f35/lhotse/cut.py#L3358-L3410

Reply:
>took too much memory

How is this problem solved?
I think originally feats are already stored in HDF5 format.

Reply:
For gigaspeech, do you have any recommand values of num_workers and batch_duration when using compute_and_store_features_batch?
Now the default values are used, i.e. num_workers=4 and batch_duration=600.0

Reply:
> > took too much memory
> 
> How is this problem solved? I think originally feats are already stored in HDF5 format.

It’s about the RAM memory, not disk usage. Basically for 1h+ waveforms naive feature extraction implementations use quadratic memory. Kaldifeat has a mechanism to prevent that. But anyway without CUDA it would have taken too much time to compute feats for 10k hours.

> For gigaspeech, do you have any recommand values of num_workers and batch_duration when using compute_and_store_features_batch?
Now the default values are used, i.e. num_workers=4 and batch_duration=600.0

It should work fine with defaults; you can try increasing these but I found that on some systems it generates too much I/O if you create too many workers and the DataLoader crashes.